### PR TITLE
Ignore calc() in length-zero-no-unit

### PR DESCRIPTION
**Pull request by vankop**
> Which issue, if any, is this issue related to?

Closes https://github.com/stylelint/stylelint/issues/3037

> Is there anything in the PR that needs further explanation?

nothing

### CODE REVIEW

File: lib/rules/length-zero-no-unit/index.js [hunk 1]
```diff
@@ -28,7 +28,11 @@ const rule = function(actual, secondary, context) {
     }
 
     root.walkDecls(decl => {
-      check(blurComments(decl.toString()), decl);
+      const stringValue = blurComments(decl.toString());
+
+      if (!/calc\((.|\s)*\)/i.test(stringValue)) {
```
Comment (evilebottnawi) [inline lib/rules/length-zero-no-unit/index.js]: Let's move this to util and name file `isMathFunction` and put logic in utils (in next css spec we will have more math functions, not only `calc`)

Comment (vankop) [reply lib/rules/length-zero-no-unit/index.js]: done

File: lib/rules/length-zero-no-unit/__tests__/index.js [hunk 2]
```diff
@@ -12,6 +12,18 @@ testRule(rule, {
...
+      code: "a { padding: calc(0px +\n 0px); }"
+    },
+    {
+      description: "ignore calc. insensitive",
+      code: "a { padding: cAlc(0px + 0px); }"
+    },
+    {
+      description: "ignore calc. empty calc",
+      code: "a { padding: calc(); }"
+    },
```
Comment (evilebottnawi) [inline lib/rules/length-zero-no-unit/__tests__/index.js]: Can we add test like `padding: calc(1in + 0in * 2)) 0px`, should have one error

Comment (vankop) [reply lib/rules/length-zero-no-unit/__tests__/index.js]: awesome suggestion. I have improve it to `padding: calc(1in + 0in * 2)) 0in calc(0px) 0px` since it was needed to rework regexp, too

File: lib/rules/length-zero-no-unit/__tests__/index.js [hunk 3]
```diff
@@ -258,6 +262,15 @@ testRule(rule, {
...
       column: 27
     },
+    {
+      description: "ignore calc. has another zero units",
+      code: "a { padding: calc(1in + 0in * 2)) 0in calc(0px) 0px }",
+      fixed: "a { padding: calc(1in + 0in * 2)) 0 calc(0px) 0 }",
+
+      message: messages.rejected,
+      line: 1,
+      column: 36
```
Comment (evilebottnawi) [inline lib/rules/length-zero-no-unit/__tests__/index.js]: And other hardcore example: `padding: calc(var(--foo, 0in) + 10px)` :smile:

File: lib/utils/getMathFunctionsRanges.js [hunk 4]
```diff
@@ -1,12 +1,16 @@
 "use strict";
 
-const MATH_REGEXP = /calc\([^()]*(?:\(.*\))*[^()]*\)/gi;
+const MATH_FUNCTIONS = ["calc"];
+const MATH_FUNCTIONS_REGEXP = new RegExp(
+  `(?:${MATH_FUNCTIONS.join("|")})` + "\\([^()]*(?:\\(.*\\))*[^()]*\\)",
+  "gi"
+);
```
Comment (evilebottnawi) [inline lib/utils/getMathFunctionsRanges.js]: Please don't use regexp for this cases, we have postcss-value-parser, regexp is slow and unsafe

Comment (vankop) [reply lib/utils/getMathFunctionsRanges.js]: 👌

File: lib/rules/length-zero-no-unit/__tests__/index.js (hunk 3 above)
Comment (vankop) [reply lib/rules/length-zero-no-unit/__tests__/index.js]: `ignore: ["custom-properties"]` does not work with such example, is it a bug?

Comment (evilebottnawi) [reply lib/rules/length-zero-no-unit/__tests__/index.js]: i think yes, we catch a bug

File: lib/rules/length-zero-no-unit/README.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
 # length-zero-no-unit
 
-Disallow units for zero lengths.
+Disallow units for zero lengths. Ignores `calc` value in favor of `function-calc-no-invalid` rule.
```
Comment (jeddy3) [inline lib/rules/length-zero-no-unit/README.md]: Minor doc nitpick. Let's revert the change to this sentence and add the following _below_ the example:

```md
This rule ignores lengths within `calc` functions in favor of the [`function-calc-no-invalid`](../function-calc-no-invalid/README.md) rule.
```

This is to be consistent with other rule READMEs e.g. 
https://github.com/stylelint/stylelint/tree/master/lib/rules/font-family-name-quotes#font-family-name-quotes


Otherwise, this PR LGTM.

Comment (vankop) [reply lib/rules/length-zero-no-unit/README.md]: updated

File: lib/rules/length-zero-no-unit/README.md [hunk 6]
```diff
@@ -1,6 +1,6 @@
-# length-zero-no-unit
+  # length-zero-no-unit
 
-Disallow units for zero lengths.
+Disallow units for zero lengths. This rule ignores lengths within math functions (e.g. `calc`) in favor of the [`function-calc-no-invalid`](../function-calc-no-invalid/README.md) rule.
```
Comment (jeddy3) [inline lib/rules/length-zero-no-unit/README.md]: Can you please move the "This rule ignores" sentence _to below the example code block_ as we use this documentation pattern in lots of other places ([like so](https://github.com/stylelint/stylelint/tree/master/lib/rules/font-family-name-quotes#font-family-name-quotes)).

Comment (vankop) [reply lib/rules/length-zero-no-unit/README.md]: Hope now is fine

File: lib/utils/isMathFunction.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+/**
+ * Check whether a node is math function
+ *
+ * @param {Node} postcss-value-parser node
+ * @return {boolean} If `true`, the node is math function
+ */
+module.exports = function isMathFunction(node /*: Object */) /*: boolean*/ {
+  return (
+    node.type === "function" &&
+    MATH_FUNCTIONS.indexOf(node.value.toLowerCase()) > -1
```
Comment (evilebottnawi) [inline lib/utils/isMathFunction.js]: We can use this `MATH_FUNCTIONS.includes(node.value.toLowerCase())`, `indexOf` is very old, we can use more readable code :smile: 

And move `MATH_FUNCTIONS` to https://github.com/stylelint/stylelint/tree/master/lib/reference for better maintenance

Comment (vankop) [reply lib/utils/isMathFunction.js]: done

File: lib/rules/length-zero-no-unit/README.md [hunk 8]
```diff
@@ -1,4 +1,4 @@
-# length-zero-no-unit
+  # length-zero-no-unit
```
Comment (hudochenkov) [inline lib/rules/length-zero-no-unit/README.md]:
```suggestion
# length-zero-no-unit
```

Comment (hudochenkov) [inline lib/rules/length-zero-no-unit/README.md]:
```suggestion
# length-zero-no-unit
```

Comment (hudochenkov) [reply lib/rules/length-zero-no-unit/README.md]: Another reason to work in stylelint repo, and not in a fork:

Screenshot 2019-08-11 at 12 15 52

I could make this change myself and merge PR. But now we need to wait for you to accept suggestion, and only then we can merge :)

Comment (vankop) [reply lib/rules/length-zero-no-unit/README.md]: fixed

Comment (vankop) [reply lib/rules/length-zero-no-unit/README.md]: > I could make this change myself and merge PR. But now we need to wait for you to accept suggestion, and only then we can merge :)

for sure the most valued advantage =)

**Comment by hudochenkov** [comment]
> -   Fixed: `length-zero-no-unit` false positives for inside calc function ([#4175](https://github.com/stylelint/stylelint/pull/4175)).
